File: platform/lang-api/src/com/intellij/codeInsight/hints/InlayParameterHintsProvider.java
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.codeInsight.hints;

import com.intellij.codeInsight.CodeInsightBundle;
import com.intellij.lang.Language;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.SyntaxTraverser;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Set;

/**
 * Provides simple text inlays (info elements rendered inside source code) for a given language.
 * The order of hints which share the same offset is not guaranteed.
 *
 * @see InlayHintsProvider for more interactive inlays.
 */
public interface InlayParameterHintsProvider {

  /**
   * @param element element for which hints should be shown.
   * @return list of hints to be shown, hints offsets should be located within element's text range.
   */
  @NotNull
  default List<InlayInfo> getParameterHints(@NotNull PsiElement element) {
    return Collections.emptyList();
  }

  /**
   * @param file file which holds element.
   * @param element element for which hints should be shown.
   * @return list of hints to be shown, hints offsets should be located within element's text range.
   */
  @NotNull
  default List<InlayInfo> getParameterHints(@NotNull PsiElement element, @NotNull PsiFile file) {
    return getParameterHints(element);
  }

  /**
   * Provides information about hint for intention actions (can be {@link HintInfo.MethodInfo} or {@link HintInfo.OptionInfo})
   * which allow enabling/disabling hints at a given position.
   *
   * Make sure that this method executed fast enough to run on EDT.
   *
   * @param element the element under the caret
   *
   */
  @Nullable
  default HintInfo getHintInfo(@NotNull PsiElement element) {
    return null;
  }

  /**
   * Provides information about hint for intention actions (can be {@link HintInfo.MethodInfo} or {@link HintInfo.OptionInfo})
   * which allow enabling/disabling hints at a given position.
   *
   * Make sure that this method executed fast enough to run on EDT.
   *
   * @param element the element under the caret
   */
  @Nullable
  default HintInfo getHintInfo(@NotNull PsiElement element, @NotNull PsiFile file) {
    return getHintInfo(element);
  }

  /**
   * Exclude list - default list of patterns for which hints should not be shown.
   */
  @NotNull
  Set<String> getDefaultBlackList();

  /**
   * Returns language which exclude list will be appended to the resulting one.
   * E.g. to prevent possible Groovy and Kotlin extensions from showing hints for excluded Java methods.
   */
  @Nullable
  default Language getBlackListDependencyLanguage() {
    return null;
  }

  /**
   * List of supported options, shown in settings dialog.
   */
  @NotNull
  default List<Option> getSupportedOptions() {
    return ContainerUtil.emptyList();
  }

  /**
   * If {@code false} no exclude list panel will be shown in "File | Settings | Editor | Inlay Hints | Language | Parameter Hints".
   */
  default boolean isBlackListSupported() {
    return true;
  }

  /**
   * Text explaining exclude list patterns.
   */
  @Nls
  default String getBlacklistExplanationHTML() {
    return CodeInsightBundle.message("inlay.hints.exclude.list.pattern.explanation");
  }

  /**
   * Customise hints presentation.
   */
  @NotNull
  default String getInlayPresentation(@NotNull String inlayText) {
    return inlayText + ":";
  }

  /**
   * Whether provider should be queried for hints ({@link #getParameterHints(PsiElement)}) even if showing hints is disabled globally.
   * ({@link com.intellij.openapi.editor.ex.EditorSettingsExternalizable#isShowParameterNameHints()}).
   */
  default boolean canShowHintsWhenDisabled() { return false; }

  /**
   * @deprecated the text is not rendered in settings anymore, do not implement this method.
   */
  @Deprecated
  default String getSettingsPreview() { return null; }

  /**
   * @return {@code true} if set of options is exhaustive and if all options are disabled, provider will collect no hints.
   */
  default boolean isExhaustive() {
    return false;
  }

  /**
   * @return text of main checkbox in hints settings.
   */
  default String getMainCheckboxText() {
    return CodeInsightBundle.message("settings.inlay.show.parameter.hints");
  }

  /**
   * @return Traverser for `root` element subtree.
   */
  @NotNull
  default SyntaxTraverser<PsiElement> createTraversal(@NotNull PsiElement root) {
    return SyntaxTraverser.psiTraverser(root);
  }

  @Nls
  default String getDescription() { return null; }

  /**
   * @param key bundle key of the option.
   * @return description of the given option or null (in this case it won't be shown).
   */
  @Nls
  @Nullable
  default String getProperty(String key) { return null; }
}

File: platform/lang-api/src/com/intellij/codeInsight/hints/InlayParameterHintsProvider.kt
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.codeInsight.hints

import com.intellij.codeInsight.CodeInsightBundle
import com.intellij.codeInsight.hints.settings.ParameterNameHintsSettings
import com.intellij.lang.Language
import com.intellij.lang.LanguageExtension
import com.intellij.lang.LanguageExtensionPoint
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.extensions.ExtensionPointName
import com.intellij.openapi.util.NlsActions
import com.intellij.openapi.util.NlsContexts
import com.intellij.psi.PsiElement
import org.jetbrains.annotations.Nls
import org.jetbrains.annotations.NonNls
import java.util.function.Supplier

val PARAMETER_NAME_HINTS_EP = ExtensionPointName.create<LanguageExtensionPoint<InlayParameterHintsProvider>>("com.intellij.codeInsight.parameterNameHints")

object InlayParameterHintsExtension : LanguageExtension<InlayParameterHintsProvider>(PARAMETER_NAME_HINTS_EP)

/**
 * If you are just implementing parameter hints, only three options are relevant to you: text, offset, isShowOnlyIfExistedBefore. The rest
 * can be used in completion hints.
 *
 * @property text hints text to show
 * @property offset offset in document where hint should be shown
 * @property isShowOnlyIfExistedBefore defines if hint should be shown only if it was present in editor before update
 *
 * @property isFilterByExcludeList allows to prevent hints from filtering by blacklist matcher (possible use in completion hints)
 * @property relatesToPrecedingText whether hint is associated with previous or following text
 * @property widthAdjustment allows resulting hint's width to match certain editor text's width, see [HintWidthAdjustment]
 */
class InlayInfo(val text: String,
                val offset: Int,
                val isShowOnlyIfExistedBefore: Boolean,
                val isFilterByExcludeList: Boolean,
                val relatesToPrecedingText: Boolean,
                val widthAdjustment: HintWidthAdjustment?) {

  constructor(text: String, offset: Int, isShowOnlyIfExistedBefore: Boolean, isFilterByBlacklist: Boolean, relatesToPrecedingText: Boolean)
    : this(text, offset, isShowOnlyIfExistedBefore, isFilterByBlacklist, relatesToPrecedingText, null)
  constructor(text: String, offset: Int, isShowOnlyIfExistedBefore: Boolean) : this(text, offset, isShowOnlyIfExistedBefore, true, false)
  constructor(text: String, offset: Int) : this(text, offset, false, true, false)

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other?.javaClass != javaClass) return false

    other as InlayInfo

    if (text != other.text) return false
    if (offset != other.offset) return false

    return true
  }

  override fun hashCode(): Int {
    var result = text.hashCode()
    result = 31 * result + offset
    return result
  }

}


sealed class HintInfo {

  /**
   * Provides fully qualified method name (e.g. "java.util.Map.put") and list of its parameter names.
   * Used to match method with exclude list, and to add method into exclude list.
   * @language in case you want to put this method into blacklist of another language.
   */
  open class MethodInfo(val fullyQualifiedName: String, val paramNames: List<String>, val language: Language?) : HintInfo() {
    constructor(fullyQualifiedName: String, paramNames: List<String>) : this(fullyQualifiedName, paramNames, null)

    /**
     * Presentable method name which will be shown to the user when adding it to exclude list.
     */
    open fun getMethodName(): String {
      val start = fullyQualifiedName.lastIndexOf('.') + 1
      return fullyQualifiedName.substring(start)
    }

    /**
     * Presentable text which will be shown in the inlay hints popup menu.
     */
    @NlsActions.ActionText
    open fun getDisableHintText(): String {
      return CodeInsightBundle.message("inlay.hints.show.settings", getMethodName())
    }
  }

  /**
   * Provides option to disable/enable by alt-enter.
   */
  open class OptionInfo(protected val option: Option) : HintInfo() {

    open fun disable(): Unit = alternate()
    open fun enable(): Unit = alternate()

    private fun alternate() {
      val current = option.get()
      option.set(!current)
    }

    open val optionName: String = option.name

    fun isOptionEnabled(): Boolean = option.isEnabled()

  }

  open fun isOwnedByPsiElement(elem: PsiElement, editor: Editor): Boolean {
    val textRange = elem.textRange
    if (textRange == null) return false
    val start = if (textRange.isEmpty) textRange.startOffset else textRange.startOffset + 1
    return editor.inlayModel.hasInlineElementsInRange(start, textRange.endOffset)
  }
}

/**
 * @param id global unique identifier of this option
 * @param nameSupplier user visible name supplier
 */
data class Option(@NonNls val id: String,
                  private val nameSupplier: Supplier<@NlsContexts.DetailedDescription String>,
                  val defaultValue: Boolean) {

  @Deprecated("Use default constructor")
  constructor(@NonNls id: String, @Nls name: String, defaultValue: Boolean) : this(id, Supplier { name }, defaultValue)

  @get:NlsContexts.DetailedDescription
  val name: String
    get() = nameSupplier.get()

  var extendedDescriptionSupplier: Supplier<@NlsContexts.DetailedDescription String>? = null

  fun isEnabled(): Boolean = get()

  fun get(): Boolean {
    return ParameterNameHintsSettings.getInstance().getOption(id) ?: defaultValue
  }

  fun set(newValue: Boolean) {
    val settings = ParameterNameHintsSettings.getInstance()
    if (newValue == defaultValue) {
      settings.setOption(id, null)
    }
    else {
      settings.setOption(id, newValue)
    }
  }

}

